### DISCUSSION TITLE
format of models and fine-tuned models (Llama2)

### DISCUSSION THREAD

**Discussion by BelloGoo** [oobabooga/text-generation-webui#3838] [Q&A]
im totally new to this, so im sorry if im asking some stupid questions.
i found that seemingly only .safetensors could be loaded as model to run, but .bin. make any sense?? if yes, only .safetensors could do, is there way to convert .bin to .safetensors? i fine-tuned an experimental model in .bin, but it couldnt be loaded.

i followed a tutorial on youtube and at the end of the fine-tuning, a model named 'adapter_model.bin' (16MB) is yielded, along with some helper files like adapter_config.json and  finetuning_args.json. i was shocked that how could the fine-tuned model with just 16MB, while the base model is a llama2-7b (daryl149/llama-2-7b-chat-hf). i start thinking if the fine-tuned model needs to couple with the base one to run? cuz i saw some mdoels on HF consist of more than one model files. 

i tried to use textgen build-in training function, but i couldnt input the dataset, even tho the dataset is written exactly the same as the  required format. the entire error message is shown below. thankyou folks.

ERROR MESSAGE:
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/gradio/routes.py", line 427, in run_predict
    output = await app.get_blocks().process_api(
  File "/usr/local/lib/python3.10/dist-packages/gradio/blocks.py", line 1323, in process_api
    result = await self.call_function(
  File "/usr/local/lib/python3.10/dist-packages/gradio/blocks.py", line 1067, in call_function
    prediction = await utils.async_iteration(iterator)
  File "/usr/local/lib/python3.10/dist-packages/gradio/utils.py", line 336, in async_iteration
    return await iterator.__anext__()
  File "/usr/local/lib/python3.10/dist-packages/gradio/utils.py", line 329, in __anext__
    return await anyio.to_thread.run_sync(
  File "/usr/local/lib/python3.10/dist-packages/anyio/to_thread.py", line 33, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/usr/local/lib/python3.10/dist-packages/anyio/_backends/_asyncio.py", line 877, in run_sync_in_worker_thread
    return await future
  File "/usr/local/lib/python3.10/dist-packages/anyio/_backends/_asyncio.py", line 807, in run
    result = context.run(func, *args)
  File "/usr/local/lib/python3.10/dist-packages/gradio/utils.py", line 312, in run_sync_iterator_async
    return next(iterator)
  File "/content/text-generation-webui/modules/training.py", line 452, in do_train
    train_data = data['train'].map(generate_and_tokenize_prompt, new_fingerprint='%030x' % random.randrange(16**30))
  File "/usr/local/lib/python3.10/dist-packages/datasets/arrow_dataset.py", line 592, in wrapper
    out: Union["Dataset", "DatasetDict"] = func(self, *args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/datasets/arrow_dataset.py", line 557, in wrapper
    out: Union["Dataset", "DatasetDict"] = func(self, *args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/datasets/arrow_dataset.py", line 3097, in map
    for rank, done, content in Dataset._map_single(**dataset_kwargs):
  File "/usr/local/lib/python3.10/dist-packages/datasets/arrow_dataset.py", line 3450, in _map_single
    example = apply_function_on_filtered_inputs(example, i, offset=offset)
  File "/usr/local/lib/python3.10/dist-packages/datasets/arrow_dataset.py", line 3353, in apply_function_on_filtered_inputs
    processed_inputs = function(*fn_args, *additional_args, **fn_kwargs)
  File "/content/text-generation-webui/modules/training.py", line 447, in generate_and_tokenize_prompt
    prompt = generate_prompt(data_point)
  File "/content/text-generation-webui/modules/training.py", line 444, in generate_prompt
    raise RuntimeError(f'Data-point "{data_point}" has no keyset match within format "{list(format_data.keys())}"')
RuntimeError: Data-point "{'instruction,input,output': 'User: %instruction%: %input%\nAssistant: %output%', 'instruction,output': 'User: Give three tips for staying healthy.\nAssistant: 1. Eat a balanced and nutritious diet: Make sure your meals are inclusive of a variety of fruits and vegetables, lean protein, whole grains, and healthy fats. This helps to provide your body with the essential nutrients to function at its best and can help prevent chronic diseases.\n\n2. Engage in regular physical activity: Exercise is crucial for maintaining strong bones, muscles, and cardiovascular health. Aim for at least 150 minutes of moderate aerobic exercise or 75 minutes of vigorous exercise each week.\n\n3. Get enough sleep: Getting enough quality sleep is crucial for physical and mental well-being. It helps to regulate mood, improve cognitive function, and supports healthy growth and immune function. Aim for 7-9 hours of sleep each night.'}" has no keyset match within format "['instruction,output', 'instruction,input,output']"

**Comment by VegaStarlake** [comment]
It sounds like your issue is almost identical to mine, but I'm not knowledgeable enough to know for sure. 

Here's my post on how I fixed it. 

https://github.com/oobabooga/text-generation-webui/issues/2710#issuecomment-1712512787

**Reply by BelloGoo** [reply]
hi, does your dataset work after you fixed it? 
i made 2 testing datasets like these. i followed the format provided in webui ([text-generation-webui/training/formats; ](https://github.com/oobabooga/text-generation-webui/tree/main/training/formats)), but still doesnt work.

**alpaca-chatbot-formatted**
`[
    {
        "instruction,output": "User: %Identify the odd one out.%\nAssistant: %The odd one out is Telegram. Twitter and Instagram are social media platforms mainly for sharing information, images and videos while Telegram is a cloud-based instant messaging and voice-over-IP service.%",
        "instruction,input,output": "User: %Identify the odd one out.%: %Twitter, Instagram, Telegram%\nAssistant: %The odd one out is Telegram. Twitter and Instagram are social media platforms mainly for sharing information, images and videos while Telegram is a cloud-based instant messaging and voice-over-IP service.%"
    },
    {
        "instruction,output": "User: %Explain why the following fraction is equivalent to 1/4%\nAssistant: %The fraction 4/16 is equivalent to 1/4 because both fractions represent the same value. A fraction can be simplified by dividing both the numerator and the denominator by a common factor. In this case, 4 is a common factor of both the numerator and the denominator of 4/16. When we divide both by 4, we get 4/4 = 1 and 16/4 = 4, so the simplified fraction is 1/4. Alternatively, we can think of this in terms of multiplication. For example, if we multiply the numerator and denominator of the fraction 1/4 by 4, we get (1x4)/(4x4), or 4/16. Since both fractions can be derived from the other through multiplication or division by the same number, they represent the same value and are equivalent.%",
        "instruction,input,output": "User: %Explain why the following fraction is equivalent to 1/4%: %4/16%\nAssistant: %The fraction 4/16 is equivalent to 1/4 because both fractions represent the same value. A fraction can be simplified by dividing both the numerator and the denominator by a common factor. In this case, 4 is a common factor of both the numerator and the denominator of 4/16. When we divide both by 4, we get 4/4 = 1 and 16/4 = 4, so the simplified fraction is 1/4. Alternatively, we can think of this in terms of multiplication. For example, if we multiply the numerator and denominator of the fraction 1/4 by 4, we get (1x4)/(4x4), or 4/16. Since both fractions can be derived from the other through multiplication or division by the same number, they represent the same value and are equivalent.%"
    },
    {
        "instruction,output": "User: %Evaluate this sentence for spelling and grammar mistakes%\nAssistant: %There are two spelling errors in the sentence. The corrected sentence should be: 'He finished his meal and left the restaurant.'%",
        "instruction,input,output": "User: %Evaluate this sentence for spelling and grammar mistakes%: %He finnished his meal and left the resturant%\nAssistant: %There are two spelling errors in the sentence. The corrected sentence should be: 'He finished his meal and left the restaurant.'%"
    },
    {
        "instruction,output": "User: %Identify the odd one out.%\nAssistant: %The odd one out is Telegram. Twitter and Instagram are social media platforms mainly for sharing information, images and videos while Telegram is a cloud-based instant messaging and voice-over-IP service.%",
        "instruction,input,output": "User: %Identify the odd one out.%: %Twitter, Instagram, Telegram%\nAssistant: %The odd one out is Telegram. Twitter and Instagram are social media platforms mainly for sharing information, images and videos while Telegram is a cloud-based instant messaging and voice-over-IP service.%"
    },
    {
        "instruction,output": "User: %Identify the odd one out.%\nAssistant: %The odd one out is Telegram. Twitter and Instagram are social media platforms mainly for sharing information, images and videos while Telegram is a cloud-based instant messaging and voice-over-IP service.%",
        "instruction,input,output": "User: %Identify the odd one out.%: %Twitter, Instagram, Telegram%\nAssistant: %The odd one out is Telegram. Twitter and Instagram are social media platforms mainly for sharing information, images and videos while Telegram is a cloud-based instant messaging and voice-over-IP service.%"
    }
    ]`
    
**alpaca-formatted**
`[
{
    "instruction,output": "\n\n### Instruction:\n%Identify the odd one out.%\n\n### Response:\n%The odd one out is Telegram. Twitter and Instagram are social media platforms mainly for sharing information, images and videos while Telegram is a cloud-based instant messaging and voice-over-IP service.%",
    "instruction,input,output": "\n\n### Instruction:\n%Identify the odd one out.%\n\n### Input:\n%Twitter, Instagram, Telegram%\n\n### Response:\n%The odd one out is Telegram. Twitter and Instagram are social media platforms mainly for sharing information, images and videos while Telegram is a cloud-based instant messaging and voice-over-IP service.%"
},
{
    "instruction,output": "\n\n### Instruction:\n%Explain why the following fraction is equivalent to 1/4%\n\n### Response:\n%The fraction 4/16 is equivalent to 1/4 because both fractions represent the same value. A fraction can be simplified by dividing both the numerator and the denominator by a common factor. In this case, 4 is a common factor of both the numerator and the denominator of 4/16. When we divide both by 4, we get 4/4 = 1 and 16/4 = 4, so the simplified fraction is 1/4. Alternatively, we can think of this in terms of multiplication. For example, if we multiply the numerator and denominator of the fraction 1/4 by 4, we get (1x4)/(4x4), or 4/16. Since both fractions can be derived from the other through multiplication or division by the same number, they represent the same value and are equivalent.%",
    "instruction,input,output": "\n\n### Instruction:\n%Explain why the following fraction is equivalent to 1/4%\n\n### Input:\n%4/16%\n\n### Response:\n%The fraction 4/16 is equivalent to 1/4 because both fractions represent the same value. A fraction can be simplified by dividing both the numerator and the denominator by a common factor. In this case, 4 is a common factor of both the numerator and the denominator of 4/16. When we divide both by 4, we get 4/4 = 1 and 16/4 = 4, so the simplified fraction is 1/4. Alternatively, we can think of this in terms of multiplication. For example, if we multiply the numerator and denominator of the fraction 1/4 by 4, we get (1x4)/(4x4), or 4/16. Since both fractions can be derived from the other through multiplication or division by the same number, they represent the same value and are equivalent.%"
},
{
    "instruction,output": "\n\n### Instruction:\n%Evaluate this sentence for spelling and grammar mistakes%\n\n### Response:\n%There are two spelling errors in the sentence. The corrected sentence should be: 'He finished his meal and left the restaurant.'%",
    "instruction,input,output": "\n\n### Instruction:\n%Evaluate this sentence for spelling and grammar mistakes%\n\n### Input:\n%He finnished his meal and left the resturant%\n\n### Response:\n%There are two spelling errors in the sentence. The corrected sentence should be: 'He finished his meal and left the restaurant.'%"
},
{
    "instruction,output": "\n\n### Instruction:\n%Identify the odd one out.%\n\n### Response:\n%The odd one out is Telegram. Twitter and Instagram are social media platforms mainly for sharing information, images and videos while Telegram is a cloud-based instant messaging and voice-over-IP service.%",
    "instruction,input,output": "\n\n### Instruction:\n%Identify the odd one out.%\n\n### Input:\n%Twitter, Instagram, Telegram%\n\n### Response:\n%The odd one out is Telegram. Twitter and Instagram are social media platforms mainly for sharing information, images and videos while Telegram is a cloud-based instant messaging and voice-over-IP service.%"
},
{
    "instruction,output": "\n\n### Instruction:\n%Identify the odd one out.%\n\n### Response:\n%The odd one out is Telegram. Twitter and Instagram are social media platforms mainly for sharing information, images and videos while Telegram is a cloud-based instant messaging and voice-over-IP service.%",
    "instruction,input,output": "\n\n### Instruction:\n%Identify the odd one out.%\n\n### Input:\n%Twitter, Instagram, Telegram%\n\n### Response:\n%The odd one out is Telegram. Twitter and Instagram are social media platforms mainly for sharing information, images and videos while Telegram is a cloud-based instant messaging and voice-over-IP service.%"
}
]
`

**Comment by FartyPants** [comment]
A common errors are often JSON errors, so you should check your dataset. The best is to see some json validator. You may have loose " or missing , or any of so many things that are hard to spot by eyes. 
Also check encoding of the json file. Best to keep it UTF-8

**Reply by BelloGoo** [reply]
![image](https://github.com/oobabooga/text-generation-webui/assets/133203509/f4c33e14-9574-4744-9975-bf040ef009fb)
   raise DatasetGenerationError("An error occurred while generating the dataset") from e
datasets.builder.DatasetGenerationError: An error occurred while generating the dataset

i validated my json and converted it to UTF-8. but i still got this error message. does it make any sense to you?
